### PR TITLE
add library classes

### PR DESCRIPTION
**Pull request by ArtemVlasenko**

### CODE REVIEW

File: Library.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class Library
+
```
Comment (differencialx) [inline Library.rb]: Why is it empty?

File: library.rb [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+  def authors_book
+    @author
+  rescue ZeroDivisionError => e
+    "argument must not be zero"
+  end
+
+  private
+
+  def validate_book
+    if @title == 0
```
Comment (differencialx) [inline library.rb]: This won't work

File: library.rb [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    @title
+  rescue ZeroDivisionError => e
+    "argument must not be zero"
+  end
+
+  def authors_book
+    @author
+  rescue ZeroDivisionError => e
+    "argument must not be zero"
+  end
```
Comment (differencialx) [inline library.rb]: Remove

Comment (differencialx) [reply library.rb]: Everywhere

File: library.rb [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+      raise ValidationError, "name must be given"
+    elsif !@title.is_a?(String)
+      raise ValidationError, "name must be string"
+    elsif @title.empty?
+      raise ValidationError, "name must be not empty"
+    end
+    if @author.nil?
+      raise ValidationError, "name must be given"
+    elsif @author.is_a?(Author)
+      raise ValidationError, "name must be string"
```
Comment (differencialx) [inline library.rb]: Please provide correct error message

File: library.rb [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+    elsif !@name.is_a?(String)
+      raise ValidationError, "name must be string"
+    elsif @name.empty?
+      raise ValidationError, "name must be not empty"
+    end
+    if !@biography.is_a?(String)
+      raise ValidationError, "name must be string"
+    elsif @biography.nil?
+      raise ValidationError, "name must be given"
+    end
```
Comment (differencialx) [inline library.rb]: Correct an order of validations

File: library.rb [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+    if @book.nil?
+      raise ValidationError, "name must be given"
+    elsif @book.is_a?(Book)
+      raise ValidationError, "name must be string"
+    end
+    if @reader.nil?
+      raise ValidationError, "name must be given"
+    elsif @reader.is_a?(Reader)
+      raise ValidationError, "name must be string"
+    end
```
Comment (differencialx) [inline library.rb:139]: Add validations for `date`

File: library.rb [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+      raise ValidationError, "name must be given"
+    elsif !@street.is_a?(String)
+      raise ValidationError, "name must be string"
+    elsif @street.empty?
+      raise ValidationError, "name must be not empty"
+    end
+    if @house.nil?
+      raise ValidationError, "name must be given"
+    elsif !@house.is_a?(Integer)
+      raise ValidationError, "name must be string"
```
Comment (differencialx) [inline library.rb]: please provide correct error message

File: library.rb [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+    "{ book: #{book}, reader: #{reader}, date: #{date} }"
+  end
+
+  private
+
+  def validate_order
+    if @book.nil?
+      raise ValidationError, "name must be given"
+    elsif @book.is_a?(Book)
+      raise ValidationError, "name must be string"
```
Comment (differencialx) [inline library.rb]: please provide correct error message

File: library.rb [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+    end
+    if @reader.nil?
+      raise ValidationError, "name must be given"
+    elsif @reader.is_a?(Reader)
+      raise ValidationError, "name must be string"
```
Comment (differencialx) [inline library.rb]: please provide correct error message

File: library.rb [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+  private
+
+  def validate_order
+    if @book.nil?
+      raise ValidationError, "name must be given"
+    elsif @book.is_a?(Book)
+      raise ValidationError, "name must be string"
+    end
+    if @reader.nil?
+      raise ValidationError, "name must be given"
```
Comment (differencialx) [inline library.rb]: please provide correct error message

File: library.rb [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+
+  def to_s
+    "{ book: #{book}, reader: #{reader}, date: #{date} }"
+  end
+
+  private
+
+  def validate_order
+    if @book.nil?
+      raise ValidationError, "name must be given"
```
Comment (differencialx) [inline library.rb]: please provide correct error message

File: library.rb [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+  def validate_autor
+    if @name.nil?
+      raise ValidationError, "name must be given"
+    elsif !@name.is_a?(String)
+      raise ValidationError, "name must be string"
+    elsif @name.empty?
+      raise ValidationError, "name must be not empty"
+    end
+    if !@biography.is_a?(String)
+      raise ValidationError, "name must be string"
```
Comment (differencialx) [inline library.rb]: please provide correct error message

File: library.rb [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+    elsif @biography.nil?
+      raise ValidationError, "name must be given"
```
Comment (differencialx) [inline library.rb]: please provide correct error message

File: library.rb [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+  def validate_reader
+    if @name.nil?
+      raise ValidationError, "name must be given"
+    elsif !@name.is_a?(String)
+      raise ValidationError, "name must be string"
+    elsif @name.empty?
+      raise ValidationError, "name must be not empty"
+    end
+    if @email.nil?
+      raise ValidationError, "name must be given"
```
Comment (differencialx) [inline library.rb]: please provide correct error message

File: library.rb [hunk 15]
```diff
@@ -0,0 +1,218 @@
...
+    elsif !@email.is_a?(String)
+      raise ValidationError, "name must be string"
```
Comment (differencialx) [inline library.rb]: please provide correct error message

File: library.rb [hunk 16]
```diff
@@ -0,0 +1,218 @@
...
+    elsif @email.empty?
+      raise ValidationError, "name must be not empty"
```
Comment (differencialx) [inline library.rb]: please provide correct error message

File: library.rb [hunk 17]
```diff
@@ -0,0 +1,218 @@
...
+    end
+    if @city.nil?
+      raise ValidationError, "name must be given"
```
Comment (differencialx) [inline library.rb]:
```suggestion
      raise ValidationError, "city must be given"
```

File: library.rb [hunk 18]
```diff
@@ -0,0 +1,218 @@
...
+    elsif !@city.is_a?(String)
+      raise ValidationError, "name must be string"
```
Comment (differencialx) [inline library.rb]: please provide correct error message

File: library.rb [hunk 19]
```diff
@@ -0,0 +1,218 @@
...
+    elsif @city.empty?
+      raise ValidationError, "name must be not empty"
```
Comment (differencialx) [inline library.rb]: please provide correct error message

File: library.rb [hunk 20]
```diff
@@ -0,0 +1,218 @@
...
+    end
+    if @street.nil?
+      raise ValidationError, "name must be given"
```
Comment (differencialx) [inline library.rb]: please provide correct error message

File: library.rb [hunk 21]
```diff
@@ -0,0 +1,218 @@
...
+    elsif !@street.is_a?(String)
+      raise ValidationError, "name must be string"
```
Comment (differencialx) [inline library.rb]: please provide correct error message

File: library.rb [hunk 22]
```diff
@@ -0,0 +1,218 @@
...
+    elsif @street.empty?
+      raise ValidationError, "name must be not empty"
```
Comment (differencialx) [inline library.rb]: please provide correct error message

File: library.rb [hunk 23]
```diff
@@ -0,0 +1,218 @@
...
+    end
+    if @house.nil?
+      raise ValidationError, "name must be given"
```
Comment (differencialx) [inline library.rb]: please provide correct error message

File: library.rb [hunk 24]
```diff
@@ -0,0 +1,218 @@
...
+    elsif !@house.is_a?(Integer)
+      raise ValidationError, "name must be string"
+    elsif !@house.positive?
+      raise ValidationError, "name must be not empty"
```
Comment (differencialx) [inline library.rb]: please provide correct error message

File: library.rb [hunk 25]
```diff
@@ -0,0 +1,218 @@
...
+  def validate_book
+    if @title.nil?
+      raise ValidationError, "name must be given"
+    elsif !@title.is_a?(String)
+      raise ValidationError, "name must be string"
+    elsif @title.empty?
+      raise ValidationError, "name must be not empty"
+    end
+    if @author.nil?
+      raise ValidationError, "name must be given"
```
Comment (differencialx) [inline library.rb]: please provide correct error message

File: library.rb [hunk 26]
```diff
@@ -0,0 +1,218 @@
...
+
+  def to_s
+    "{ title: #{title}, author: #{author} }"
+  end
+
+  private
+
+  def validate_book
+    if @title.nil?
+      raise ValidationError, "name must be given"
```
Comment (differencialx) [inline library.rb]: please provide correct error message

File: library.rb [hunk 27]
```diff
@@ -0,0 +1,218 @@
...
+    elsif !@title.is_a?(String)
+      raise ValidationError, "name must be string"
```
Comment (differencialx) [inline library.rb]: please provide correct error message

File: library.rb [hunk 28]
```diff
@@ -0,0 +1,218 @@
...
+    elsif @title.empty?
+      raise ValidationError, "name must be not empty"
```
Comment (differencialx) [inline library.rb]: please provide correct error message

File: library.rb [hunk 29]
```diff
@@ -0,0 +1,218 @@
...
+reader = Reader.new('name1', 'email1', 'city1', 'street1', 1)
+order = Order.new('0', 0)
+library.add_book(book)
+library.add_author(author)
+library.add_order(order)
+library.add_reader(reader)
+# author_error_nil = Author.new(0, '2')
+# author_error_class = Author.new()
+# author_error_empty = Author.new()
+#<class_name_downcase>_<atrribute>_<validation>
```
Comment (differencialx) [inline library.rb]: Add examples with `Book`, `Reader`, `Author`, `Order` creation for cases when validation error will be raised e.g.

if validation error raised for cases when name is not provided, then you should call varialble like `author_name_nil`
```ruby
author_name_nil = Author.new(nil, 'biography')
#OR
author_biography_class = Author.new('Name', Array.new)
```

All this examples should be commented out, to make testing easier.

**Review by differencialx — CHANGES_REQUESTED** [review]

File: library.rb (hunk 4 above)
Comment (ArtemVlasenko) [reply library.rb]: done [here](https://github.com/ArtemVlasenko/Library/pull/1/commits/1f19a633e20dade406473e5d2a2d57e13fd8ad34)

File: library.rb (hunk 27 above)
Comment (ArtemVlasenko) [reply library.rb]: done [here](https://github.com/ArtemVlasenko/Library/pull/1/commits/d815a3744626f403f3f3cf72cc781d0bde4ffa24)

File: library.rb [hunk 30]
```diff
@@ -0,0 +1,218 @@
...
+  end
+
+  def to_s
+    "Books:\n #{@books.map(&:to_s).join("\n")}\n"\
+    "Authors:\n #{@authors.map(&:to_s).join("\n")}\n"\
+    "Orders:\n #{@orders.map(&:to_s).join("\n")}\n"\
+    "Readers:\n #{@readers.map(&:to_s).join("\n")}"
+  end
+
+  def save_from_file
```
Comment (differencialx) [inline library.rb]: rename `save_to_file`

File: library.rb [hunk 31]
```diff
@@ -0,0 +1,218 @@
...
+    File.open("to_s.txt", "w+") do |f|
```
Comment (differencialx) [inline library.rb]: create a constant with file name where library should be saved

File: library.rb [hunk 32]
```diff
@@ -0,0 +1,218 @@
...
+      @books.each { |element| f.puts "Book: #{(element)}" }
+      @authors.each { |element| f.puts "Author: #{(element)}" }
+      @orders.each { |element| f.puts "Order: #{(element)}" }
+      @readers.each { |element| f.puts "Readers: #{(element)}" }
+    end
+  end
+
+  def read_from_file
+    file = File.readlines("to_s.txt")
```
Comment (differencialx) [inline library.rb]: Use file name constant

File: library.rb (hunk 31 above)
Comment (differencialx) [inline library.rb]: what `w+` means?

File: library.rb [hunk 33]
```diff
@@ -0,0 +1,218 @@
...
+  def to_s
+    "Books:\n #{@books.map(&:to_s).join("\n")}\n"\
+    "Authors:\n #{@authors.map(&:to_s).join("\n")}\n"\
+    "Orders:\n #{@orders.map(&:to_s).join("\n")}\n"\
+    "Readers:\n #{@readers.map(&:to_s).join("\n")}"
+  end
+
+  def save_from_file
+    File.open("to_s.txt", "w+") do |f|
+      @books.each { |element| f.puts "Book: #{(element)}" }
```
Comment (differencialx) [inline library.rb]: 1. Create string which should written to file. Use `YAML::dump`
2. Write string from above to file

All steps above should be applied to `library`

File: library.rb [hunk 34]
```diff
@@ -0,0 +1,218 @@
+require "date"
+
+class Library
+  attr_reader :books, :authors
+  def initialize
+    @books = []
+    @authors = []
+    @orders = []
+    @readers = []
+
```
Comment (differencialx) [inline library.rb:17]: # rubocop!!!

File: library.rb [hunk 35]
```diff
@@ -0,0 +1,218 @@
+require "date"
+
+class Library
+  attr_reader :books, :authors
```
Comment (differencialx) [inline library.rb]: add readers and orders

File: library.rb [hunk 36]
```diff
@@ -0,0 +1,218 @@
...
+      @books.each { |element| f.puts "Book: #{(element)}" }
+      @authors.each { |element| f.puts "Author: #{(element)}" }
+      @orders.each { |element| f.puts "Order: #{(element)}" }
+      @readers.each { |element| f.puts "Readers: #{(element)}" }
+    end
+  end
+
+  def read_from_file
+    file = File.readlines("to_s.txt")
+    # file.map! {|i| i.chomp}
```
Comment (differencialx) [inline library.rb]: YAML::load(<serialized_data>)

Comment (differencialx) [reply library.rb]: This method should return an instance of deserialized `Library`

File: library.rb (hunk 13 above)
Comment (ArtemVlasenko) [reply library.rb]: done [here](https://github.com/ArtemVlasenko/Library/pull/1/commits/2a7d44b9fa3d020bc48fd5e204be17dc9802b1f0)

File: library.rb (hunk 6 above)
Comment (ArtemVlasenko) [inline library.rb]: done [here](https://github.com/ArtemVlasenko/Library/pull/1/commits/809f17ebcf158b912ce3e7e1e497959909b5d990)

File: library.rb (hunk 8 above)
Comment (ArtemVlasenko) [reply library.rb]: done [here](https://github.com/ArtemVlasenko/Library/pull/1/commits/809f17ebcf158b912ce3e7e1e497959909b5d990)

File: library.rb (hunk 9 above)
Comment (ArtemVlasenko) [reply library.rb]: done [here](https://github.com/ArtemVlasenko/Library/pull/1/commits/809f17ebcf158b912ce3e7e1e497959909b5d990)

File: library.rb (hunk 25 above)
Comment (ArtemVlasenko) [reply library.rb]: done [here](https://github.com/ArtemVlasenko/Library/pull/1/commits/df142999c90c153131d0f42535e819ebbc9373e1)

File: library.rb (hunk 26 above)
Comment (ArtemVlasenko) [reply library.rb]: done [here](https://github.com/ArtemVlasenko/Library/pull/1/commits/b92df4c380d81bf57debbe760cdf73234825652a)

File: library.rb (hunk 28 above)
Comment (ArtemVlasenko) [reply library.rb]: done [here](https://github.com/ArtemVlasenko/Library/pull/1/commits/b92df4c380d81bf57debbe760cdf73234825652a)

File: library.rb (hunk 10 above)
Comment (ArtemVlasenko) [reply library.rb]: done [here](https://github.com/ArtemVlasenko/Library/pull/1/commits/809f17ebcf158b912ce3e7e1e497959909b5d990)

File: library.rb (hunk 11 above)
Comment (ArtemVlasenko) [reply library.rb]: done [here](https://github.com/ArtemVlasenko/Library/pull/1/commits/809f17ebcf158b912ce3e7e1e497959909b5d990)

File: library.rb (hunk 12 above)
Comment (ArtemVlasenko) [reply library.rb]: done [here](https://github.com/ArtemVlasenko/Library/pull/1/commits/7863909e7bf508c41f7c72a8e731034354f06bb5)

File: library.rb (hunk 14 above)
Comment (ArtemVlasenko) [reply library.rb]: done [here](https://github.com/ArtemVlasenko/Library/pull/1/commits/bc3688f349677ba8303487dec53077829fc7fb89)

File: library.rb (hunk 15 above)
Comment (ArtemVlasenko) [reply library.rb]: done [here](https://github.com/ArtemVlasenko/Library/pull/1/commits/bc3688f349677ba8303487dec53077829fc7fb89)

File: library.rb (hunk 16 above)
Comment (ArtemVlasenko) [reply library.rb]: done [here](https://github.com/ArtemVlasenko/Library/pull/1/commits/bc3688f349677ba8303487dec53077829fc7fb89)

File: library.rb (hunk 18 above)
Comment (ArtemVlasenko) [reply library.rb]: done [here](https://github.com/ArtemVlasenko/Library/pull/1/commits/aa3192c43901c9a0907c92d69e22ecf4e96093f6)

File: library.rb (hunk 19 above)
Comment (ArtemVlasenko) [reply library.rb]: done[here](https://github.com/ArtemVlasenko/Library/pull/1/commits/aa3192c43901c9a0907c92d69e22ecf4e96093f6)

File: library.rb (hunk 20 above)
Comment (ArtemVlasenko) [reply library.rb]: done [here](https://github.com/ArtemVlasenko/Library/pull/1/commits/bb22afd09eba01fcf96730eadbfdfbbc45cf85d2)

File: library.rb (hunk 21 above)
Comment (ArtemVlasenko) [reply library.rb]: done [here](https://github.com/ArtemVlasenko/Library/pull/1/commits/bb22afd09eba01fcf96730eadbfdfbbc45cf85d2)

File: library.rb (hunk 23 above)
Comment (ArtemVlasenko) [reply library.rb]: done [here](https://github.com/ArtemVlasenko/Library/pull/1/commits/e31afced110230e89f73aec7ddd6a75dddd6e406)

File: library.rb (hunk 22 above)
Comment (ArtemVlasenko) [reply library.rb]: done [here](https://github.com/ArtemVlasenko/Library/pull/1/commits/bb22afd09eba01fcf96730eadbfdfbbc45cf85d2)

File: library.rb (hunk 24 above)
Comment (ArtemVlasenko) [reply library.rb]: done [here](https://github.com/ArtemVlasenko/Library/pull/1/commits/e31afced110230e89f73aec7ddd6a75dddd6e406)

File: library.rb (hunk 30 above)
Comment (ArtemVlasenko) [reply library.rb]: done [here](https://github.com/ArtemVlasenko/Library/pull/1/commits/fbf7036f3e6b6f42177f25ab33bae2c0bcb2c803)

File: library.rb (hunk 34 above)
Comment (ArtemVlasenko) [reply library.rb:17]: done [here](https://github.com/ArtemVlasenko/Library/pull/1/commits/4c26a4a7638429fcaf62b50cf400e598705c9723)

File: library.rb (hunk 35 above)
Comment (ArtemVlasenko) [reply library.rb]: done [here](https://github.com/ArtemVlasenko/Library/pull/1/commits/feffaa41e6007f75d1e75332be8f44f4148d5fc3)

File: library.rb (hunk 31 above)
Comment (ArtemVlasenko) [reply library.rb]: done [here](https://github.com/ArtemVlasenko/Library/pull/1/commits/7393ff7c29112fbdbc98ec99ca416aa8d3eccbad)

File: library.rb (hunk 33 above)
Comment (ArtemVlasenko) [reply library.rb]: done [here](https://github.com/ArtemVlasenko/Library/pull/1/commits/c3eeb28cf2ab6163bb7bec8a55fe56d37b1e3049)

File: library.rb (hunk 11 above)
Comment (differencialx) [reply library.rb]: It's not done.

File: library.rb (hunk 10 above)
Comment (differencialx) [reply library.rb]: It's not done.

File: library.rb [hunk 37]
```diff
@@ -0,0 +1,235 @@
...
+      raise ValidationError, "book must be an instance of Book"
+    end
+    if @reader.nil?
+      raise ValidationError, "name must be given"
+    elsif @reader.is_a?(Reader)
+      raise ValidationError, "reader must be an instance of Reader"
+    end
+    if @date.nil?
+      raise ValidationError, "name must be given"
+    elsif @date.is_a?(String)
```
Comment (differencialx) [inline library.rb]: incorrect validation, please fix this

File: library.rb [hunk 38]
```diff
@@ -0,0 +1,235 @@
...
+    elsif @book.is_a?(Book)
+      raise ValidationError, "book must be an instance of Book"
+    end
+    if @reader.nil?
+      raise ValidationError, "name must be given"
+    elsif @reader.is_a?(Reader)
+      raise ValidationError, "reader must be an instance of Reader"
+    end
+    if @date.nil?
+      raise ValidationError, "name must be given"
```
Comment (differencialx) [inline library.rb]: correct error message

File: library.rb [hunk 39]
```diff
@@ -0,0 +1,235 @@
...
+      raise ValidationError, "street must be string"
+    elsif @street.empty?
+      raise ValidationError, "street must be not empty"
+    end
+    if @house.nil?
+      raise ValidationError, "house must be given"
+    elsif !@house.is_a?(Integer)
+      raise ValidationError, "house must be integer"
+    elsif !@house.positive?
+      raise ValidationError, "house must be not empty"
```
Comment (differencialx) [inline library.rb]: correct error message

File: library.rb (hunk 29 above)
Comment (differencialx) [reply library.rb]:
```ruby
begin
  puts '**************************************'
  puts "Test validation for book title is nil"
  Book.new('nil', author)
  puts "FAIL"
rescue ValidationError => error
  result = error.message == "title must be given"
  puts result ? 'PASS' : 'FAIL'

  puts '*****************************************'
end
```

File: library.rb [hunk 40]
```diff
@@ -0,0 +1,618 @@
...
+    "{ book: #{book}, reader: #{reader}, date: #{date} }"
+  end
+
+  private
+
+  def validate_order
+    if @book.nil?
+      raise ValidationError, "book must be given"
+    elsif @book.is_a?(Book)
+      raise ValidationError, "book must be an instance of Book"
```
Comment (differencialx) [inline library.rb]: fix validation

File: library.rb [hunk 41]
```diff
@@ -0,0 +1,618 @@
...
+    end
+    if @reader.nil?
+      raise ValidationError, "reader must be given"
+    elsif @reader.is_a?(Reader)
```
Comment (differencialx) [inline library.rb]: fix validation

File: library.rb [hunk 42]
```diff
@@ -0,0 +1,618 @@
...
+library.add_reader(reader)
+
+# output = File.new('to_s.yml', 'w')
+# output.puts YAML.dump(library)
+# output.close
+
+# author_name_nil = Author.new('nil', 'biography')
+# author_biography_class = Author.new('Name', String.new)
+
+# begin
```
Comment (differencialx) [inline library.rb]: try this 
```ruby
def check_validation(title, expected_error_message, &block)
  print title
  block.call
  puts " -> FAIL!!!"
rescue ValidationError => error
  result = error.message == expected_error_message
  puts result ? ' -> Pass' : ' -> FAIL!!!'
end

puts '**************************************'
check_validation("Test validation for book title is nil", "title must be given") do
  Book.new(nil, author)
end

check_validation("Test validation for book title is String", "title must be string") do
  Book.new(Array.new, author)
end
puts '*****************************************'
```

File: library.rb [hunk 43]
```diff
@@ -0,0 +1,618 @@
+require 'date'
+require 'yaml'
+
+class ValidationError < StandardError
+end
+
+class Library
+
+  SAVE_LIBRARY = 'librarys.txt'.freeze
```
Comment (differencialx) [inline library.rb]: typo

File: library.rb [hunk 44]
```diff
@@ -0,0 +1,618 @@
...
+
+  def to_s
+    "Books:\n #{@books.map(&:to_s).join("\n")}\n"\
+    "Authors:\n #{@authors.map(&:to_s).join("\n")}\n"\
+    "Orders:\n #{@orders.map(&:to_s).join("\n")}\n"\
+    "Readers:\n #{@readers.map(&:to_s).join("\n")}"
+  end
+
+  def save_to_file
+    File.open(SAVE_LIBRARY, 'w') do |file|
```
Comment (differencialx) [inline library.rb]:
```ruby
File.open(SAVE_LIBRARY, 'w') do |file|
  file.write YAML.dump(self)
end
```

File: library.rb (hunk 43 above)
Comment (differencialx) [reply library.rb]:
```suggestion
  SAVE_LIBRARY = 'librarys.yml'.freeze
```

File: library.rb [hunk 45]
```diff
@@ -0,0 +1,618 @@
...
+
+class ValidationError < StandardError
+end
+
+class Library
+
+  SAVE_LIBRARY = 'librarys.txt'.freeze
+  attr_reader :books, :authors, :orders, :readers
+
+  def initialize
```
Comment (differencialx) [inline library.rb:12]:
```ruby
  def initialize
    library = read_from_file
    @books = library&.books || []
    @authors = library&.authors || []
    @orders = library&.orders || []
    @readers = library&.readers || []
  end
```

File: library.rb [hunk 46]
```diff
@@ -0,0 +1,618 @@
...
+  def save_to_file
+    File.open(SAVE_LIBRARY, 'w') do |file|
+      file.puts YAML.dump(self)
+    end
+  end
+
+  def read_from_file
+    file = File.read(SAVE_LIBRARY)
+    file = YAML.dump(file)
+    YAML.load(file)
```
Comment (differencialx) [inline library.rb:51]: remove

File: library.rb [hunk 47]
```diff
@@ -0,0 +1,618 @@
...
+
+  def save_to_file
+    File.open(SAVE_LIBRARY, 'w') do |file|
+      file.puts YAML.dump(self)
+    end
+  end
+
+  def read_from_file
+    file = File.read(SAVE_LIBRARY)
+    file = YAML.dump(file)
```
Comment (differencialx) [inline library.rb]: remove

File: library.rb [hunk 48]
```diff
@@ -0,0 +1,618 @@
...
+    YAML.load(file)
+    puts File.exists?(SAVE_LIBRARY)
+    return unless File.exists?(SAVE_LIBRARY)
```
Comment (differencialx) [inline library.rb]: remove

File: library.rb [hunk 49]
```diff
@@ -0,0 +1,618 @@
...
+# book_error_class = Book.new()
+# book_error_empty = Book.new()
+# reader_error_nil = Reader.new()
+# reader_error_class = Reader.new()
+# reader_error_empty = Reader.new()
+# order_error_nil = Order.new()
+# order_error_class = Order.new()
+# order_error_empty = Order.new()
+#<class_name_downcase>_<atrribute>_<validation>
+library.read_from_file
```
Comment (differencialx) [inline library.rb]: remove

File: library.rb [hunk 50]
```diff
@@ -0,0 +1,479 @@
...
+# check_validation("Test validation for book title is nil", "title must be given") do
+#   Book.new(nil, author)
+# end
+
+# check_validation("Test validation for book title is String", "title must be string") do
+#   Book.new(Array.new, author)
+# end
+
+# check_validation("Test validation for book title is empty", "title must be not empty") do
+#   Book.new('', author)
```
Comment (differencialx) [inline library.rb:241]: incorrect check, fix it

File: library.rb [hunk 51]
```diff
@@ -0,0 +1,479 @@
...
+#   puts result ? ' -> Pass' : ' -> FAIL!!!'
+# end
+
+# puts '**************************************'
+# check_validation("Test validation for book author is nil", "author must be given") do
+#   Book.new('title', nil)
+# end
+
+# check_validation("Test validation for book author instance of Author", "author must be an instance of Author") do
+#   Book.new('title', author)
```
Comment (differencialx) [inline library.rb]: incorrect check, fix it

File: library.rb [hunk 52]
```diff
@@ -0,0 +1,479 @@
...
+
+  def validate_autor
+    if @name.nil?
+      raise ValidationError, "name must be given"
+    elsif !@name.is_a?(String)
+      raise ValidationError, "name must be string"
+    elsif @name.empty?
+      raise ValidationError, "name must be not empty"
+    end
+    if @biography.nil?
```
Comment (differencialx) [inline library.rb]: fix validation, biography is optional

File: library.rb (hunk 40 above)
Comment (differencialx) [reply library.rb]: fix validation

File: library.rb (hunk 41 above)
Comment (differencialx) [reply library.rb]: fix validation

File: library.rb [hunk 53]
```diff
@@ -0,0 +1,479 @@
...
+# def check_validation(order, expected_error_message, &block)
+#   print order
+#   block.call
+#   puts " -> FAIL!!!"
+# rescue ValidationError => error
+#   result = error.message == expected_error_message
+#   puts result ? ' -> Pass' : ' -> FAIL!!!'
+# end
+
+# puts '**************************************'
```
Comment (differencialx) [inline library.rb]: redo validation check for order, thay are wrong

File: library.rb [hunk 54]
```diff
@@ -0,0 +1,479 @@
...
+# check_validation("Test validation for order reader is nil", "reader must be given") do
+#   Order.new('book', nil)
+# end
+
+# check_validation("Test validation for order reader instance of Reader", "reader must be an instance of Reader") do
+#   Order.new('book', reader)
+# end
+# puts '*****************************************'
+
+# def check_validation(order, expected_error_message, &block)
```
Comment (differencialx) [inline library.rb]: remove all extra method definitions

File: library.rb [hunk 55]
```diff
@@ -0,0 +1,479 @@
...
+#   result = error.message == expected_error_message
+#   puts result ? ' -> Pass' : ' -> FAIL!!!'
+# end
+
+# puts '**************************************'
+# check_validation("Test validation for reader house is nil", "house must be given") do
+#   Reader.new('name', 'email', 'city', 'street', nil)
+# end
+
+# check_validation("Test validation for reader house is String", "house must be integer") do
```
Comment (differencialx) [inline library.rb]: fix message

File: class_author.rb [hunk 56]
```diff
@@ -0,0 +1,26 @@
+module ClassAuthor
```
Comment (differencialx) [inline class_author.rb]: remove module, rename file

File: class_book.rb [hunk 57]
```diff
@@ -0,0 +1,26 @@
+module ClassBook
```
Comment (differencialx) [inline class_book.rb]: remove module, rename file

File: class_order.rb [hunk 58]
```diff
@@ -0,0 +1,30 @@
+module ClassOrder
```
Comment (differencialx) [inline class_order.rb]: remove module, rename file

File: class_reader.rb [hunk 59]
```diff
@@ -0,0 +1,42 @@
+module ClassReader
```
Comment (differencialx) [inline class_reader.rb]: remove module, rename file

File: tests.rb [hunk 60]
```diff
@@ -0,0 +1,167 @@
...
+  result = error.message == expected_error_message
+  puts result ? ' -> Pass' : ' -> FAIL!!!'
+end
+
+def run_validation_tests
+  library = Library.new
+  author = Author.new('10', '1')
+  book = Book.new('1', '1')
+  reader = Reader.new('name1', 'email1', 'city1', 'street1', 1)
+  order = Order.new('0', '0')
```
Comment (differencialx) [inline tests.rb]: entities are creating with wrong params, please fix it.

File: tests.rb [hunk 61]
```diff
@@ -0,0 +1,167 @@
...
+    Author.new('name', Array.new)
+  end
+
+  check_validation("Test validation for author biography is empty", "biography must be not empty") do
+    Author.new('name', '')
+  end
+  puts '*****************************************'
+
+  puts '**************************************'
+  check_validation("Test validation for order book is nil", "book must be given") do
```
Comment (differencialx) [inline tests.rb:69]: Fix order validation tests, they a wrong. You should pass correct params for each test case.

File: tests.rb [hunk 62]
```diff
@@ -0,0 +1,167 @@
...
+  end
+
+  check_validation("Test validation for reader city is empty", "city must be not empty") do
+    Reader.new('name', 'email', '', 'street', 'house')
+  end
+  puts '*****************************************'
+
+  puts '**************************************'
+  check_validation("Test validation for reader street is nil", "street must be given") do
+    Reader.new('name', 'email', 'city', nil, 'house')
```
Comment (differencialx) [inline tests.rb]: house should be integer

File: module_validation.rb [hunk 63]
```diff
@@ -0,0 +1,25 @@
+module ModuleValidation
+  def validate_presence(value, error_message)
+    return if !value.nil?
+
+    raise ValidationError, error_message
+  end
+
+  def validate_string(value, error_message)
```
Comment (differencialx) [inline module_validation.rb]: remove this method, use `validate_klass` instead

File: module_validation.rb [hunk 64]
```diff
@@ -0,0 +1,25 @@
+module ModuleValidation
```
Comment (differencialx) [inline module_validation.rb:1]: add validation method for `positive`

File: library.rb [hunk 65]
```diff
@@ -0,0 +1,104 @@
+require 'date'
+require 'yaml'
+require 'pry'
+require_relative 'module_validation'
+require_relative 'tests'
+require_relative 'class_book'
```
Comment (differencialx) [inline library.rb]: use renamed file names

File: library.rb [hunk 66]
```diff
@@ -0,0 +1,104 @@
...
+
+    file_read = File.read(SAVE_LIBRARY)
+    YAML.load(file_read)
+  end
+end
+
+include ClassBook
+include ClassAuthor
+include ClassOrder
+include ClassReader
```
Comment (differencialx) [inline library.rb]: remove

File: reader.rb [hunk 67]
```diff
@@ -0,0 +1,40 @@
...
+      validate_presence(@city, "city must be given")
+      validate_klass(@city, "city must be string", String)
+      validate_empty(@city, "city must be not empty")
+      validate_presence(@street, "street must be given")
+      validate_klass(@street,"street must be string", String)
+      validate_empty(@street, "street must be not empty")
+      validate_presence(@house, "house must be given")
+      validate_klass(@house,"house must be integer", Integer)
+      validate_positive(@house, "house must be positive")
+      # if !@house.positive?
```
Comment (differencialx) [inline reader.rb]: remove commented code

File: tests.rb [hunk 68]
```diff
@@ -0,0 +1,169 @@
...
+  end
+
+  check_validation("Test validation for author biography is empty", "biography must be not empty") do
+    Author.new('name', '')
+  end
+  puts '*****************************************'
+
+  puts '**************************************'
+  check_validation("Test validation for order book is nil", "book must be given") do
+    Order.new(nil, 'reader')
```
Comment (differencialx) [inline tests.rb]: pass the `reader`, not a string

**Pull request merged**